modules/telemetry/src/measure.cpp

```diff
@@ -0,0 +1,72 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+#include "hephaestus/telemetry/measure.h"
+
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include <absl/base/thread_annotations.h>
+
+#include "hephaestus/concurrency/message_queue_consumer.h"
+#include "hephaestus/telemetry/measure_sink.h"
+
+namespace heph::telemetry {
+
+class Measure {
+public:
+  explicit Measure();
+  static void registerSink(std::unique_ptr<IMeasureSink> sink);
+
+  static void measure(const MeasureEntry& measure_entry);
+
+private:
+  [[nodiscard]] static auto instance() -> Measure&;
+  void processMeasureEntries(const MeasureEntry& entry);
+
+private:
+  std::mutex sink_mutex_;
+  std::vector<std::unique_ptr<IMeasureSink>> sinks_ ABSL_GUARDED_BY(sink_mutex_);
+  concurrency::MessageQueueConsumer<MeasureEntry> measure_entries_consumer_;
+};
```

Comment:
The singleton design is very lighweight and elegant. It does however come with the downside of repeating the component name over and over again, with the ability to create different components with typos.

From a user perspective, `Measure` could be a member variable of a class, with `component` either being a parameter of filled at compile time. The call would then be along the lines of `measure_.measure(`tag`, `json_values`)` and the timestamp automatically deduced.

If we fix the `component` at compile time however, we lose the ability to reuse modules and associate them with the module which uses them. Example: ICP can be used by component `SLAM`, but also by component `Perception` or `LoadCarrierDetection`.

Reply:
Not sure we are fully aligned here yet. If a low level module, like ICP, produces telemetry output, we would like to associate this output via the component with it's caller, e.g. SLAM or LcDetection. The tag would be "icp" in both cases. How would you implement this given the current design, where `record` specifies the component.

One solution would be to add a parent_component member to low level modules, this would however mix the module definition with telemetry specific concerns.

Reply:
The ways I see:
* ICP returns detailed information which are recorded by the caller
* Upon creation ICP is passed with the component name which uses to record information
  * This would solve also the logging 
* We make telemetry not singleton and we need to pass the object everywhere (nightmare)

Reply:
I would rule out option 2, as it mixes concerns, as discussed above.

Option 1 might be the preferable solution. It means that we need to be a bit more thoughtful on what we want to log, but that is not necessarily a bad thing. By not providing a simple "record' method to be used in any low level module without care, we encourage to not record unneeded information. I would vote for this one, given the options we have.

Regarding option 3, this indeed seems like a chore. 

Feel free to solve this discussion using option 1.

Reply:
With the current design if ICP has a function like:
```
auto solve(...) -> Result;
struct Result {
  Pose3 pose;
  Metrics metrics
};
```
in the caller module I can do:
```
auto icp_result = icp->solve(...);
heph::telemetry::record("LIDARSlam", "localization", icp_result.metrics).
```

Reply:
I have the feeling, I do not fully understand the problem with the component labeling.
Is it that @lhruby wants to track where the function that logs the metric is used?
if yes this could either be solved by structure like @filippobrizzi suggested (the metric is bubbled up to a distinct component and the tag gives the specifics) or automatically with a stacktrace (however this could become costly and we would think how we would do the caching of the stacktrace so we don't evaluate it each time).
However tot me this sounds like it has some overlap with the "tracing" ...

Reply:
In order to avoid typos, we could use StrongTypes instead of strings to describe the component.
that can be converted to String inside of record(). This would be an optional interface, since the component also might be dynamically determined.

Reply:
these are all great points. I suggest we get this in and then we consider a refactor when we have more cases
otherwise we risk spinning around forever